File: src/jvmception/objects/cp/CpUtf8.java
package jvmception.objects.cp;

import java.io.DataInputStream;
import java.io.IOException;

import jvmception.jvmtypes.IUnitSerializable;
import jvmception.jvmtypes.JVMReference;
import jvmception.objects.JVMInstance;
 
public class CpUtf8 extends CpInfo {
	private String str;
	private JVMReference reference;
	
	public CpUtf8(DataInputStream dis) throws IOException {
		int lenght = dis.readUnsignedShort();
		byte[] data = dis.readNBytes(lenght);
		str = new String(data);
	}
	
	public String getString() {
		return str;
	}
	
	@Override
	public String toString() {
		return str;
	}

	private JVMReference getStringReference() {
		if (reference != null) {
			return reference;
		}
		try {
			JVMInstance instance = new JVMInstance(JVMClassFileLoader.getClassByName("java/lang/String"));
			/* TODO load string or something idk */
			return instance.addReference();
		} catch (ClassNotFoundException e) {
			e.printStackTrace();
			return null;
		}
	}
	
	@Override
	public IUnitSerializable toIUnitSerializable() {
		return reference;
	}
}

File: src/jvmception/objects/cp/CpUtf8.cs
using System;
using System.IO;

public class CpUtf8 : CpInfo
{
    private string str;
    private JVMReference reference;

    public CpUtf8(BinaryReaderBigEndian dis)
    {
        int lenght = dis.ReadUInt16();
        byte[] data = dis.ReadBytes(lenght);
        str = System.Text.Encoding.ASCII.GetString(data);
    }

    public string getString()
    {
        return str;
    }

    public override string ToString()
    {
        return str;
    }

    private JVMReference getStringReference()
    {
        if (reference != null)
        {
            return reference;
        }
        try
        {
            JVMInstance instance = new JVMInstance(JVMClassFileLoader.getClassByName("java/lang/string"), null);
            /* TODO load string or something idk */
            return instance.createReference();
        }
        catch (Exception e)
        {
            Console.WriteLine(e);
            return null;
        }
    }

    public override IUnitSerializable toIUnitSerializable()
    {
        return reference;
    }
}
